x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/DataTiersUsageTransportAction.java

```diff
@@ -0,0 +1,146 @@
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.store.StoreStats;
+import org.elasticsearch.protocol.xpack.XPackUsageRequest;
+import org.elasticsearch.search.aggregations.metrics.TDigestState;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.action.XPackUsageFeatureAction;
+import org.elasticsearch.xpack.core.action.XPackUsageFeatureResponse;
+import org.elasticsearch.xpack.core.action.XPackUsageFeatureTransportAction;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+
+public class DataTiersUsageTransportAction extends XPackUsageFeatureTransportAction {
+
+    private final Client client;
+
+    @Inject
+    public DataTiersUsageTransportAction(TransportService transportService, ClusterService clusterService,
+                                         ThreadPool threadPool, ActionFilters actionFilters,
+                                         IndexNameExpressionResolver indexNameExpressionResolver, Client client) {
+        super(XPackUsageFeatureAction.DATA_TIERS.name(), transportService, clusterService,
+            threadPool, actionFilters, indexNameExpressionResolver);
+        this.client = client;
+    }
+
+    @Override
+    protected void masterOperation(Task task, XPackUsageRequest request, ClusterState state,
+                                   ActionListener<XPackUsageFeatureResponse> listener) {
+        client.admin().cluster().prepareNodesStats()
+            .all()
+            .setIndices(CommonStatsFlags.ALL)
+            .execute(ActionListener.wrap(nodesStatsResponse -> {
+                final RoutingNodes routingNodes = state.getRoutingNodes();
+
+                // First separate the nodes into separate tiers, note that nodes *may* be duplicated
+                Map<String, List<NodeStats>> tierSpecificNodeStats = separateTiers(nodesStatsResponse);
+
+                // Generate tier specific stats for the nodes
+                Map<String, DataTiersFeatureSetUsage.TierSpecificStats> tierSpecificStats = tierSpecificNodeStats.entrySet()
+                    .stream().collect(Collectors.toMap(Map.Entry::getKey, ns -> calculateStats(ns.getValue(), routingNodes)));
+
+                listener.onResponse(new XPackUsageFeatureResponse(new DataTiersFeatureSetUsage(tierSpecificStats)));
+            }, listener::onFailure));
+    }
+
+    private static Map<String, List<NodeStats>> separateTiers(NodesStatsResponse nodesStatsResponse) {
```

Comment:
shall we unit test these static methods?

Reply:
I added a unit test for this one (and the MAD calculation), the larger one would require so much boilerplate and/or mocking as to be pretty useless as a unit test (hence the integration test for it)